docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook.go

```diff
@@ -0,0 +1,187 @@
+// +kubebuilder:webhook:path=/validate-batch-tutorial-kubebuilder-io-v1-cronjob,mutating=false,failurePolicy=fail,groups=batch.tutorial.kubebuilder.io,resources=cronjobs,verbs=create;update,versions=v1,name=vcronjob.kb.io
+
+/*
+We use the `webhook.Defaulter` interface to set defaults to our CRD.
+A webhook will automatically be served that calls this defaulting.
+*/
+
+var _ webhook.Defaulter = &CronJob{}
+
+/*
+The Default method is expected to mutate the receiver, setting the defaults.
+*/
+func (c *CronJob) Default() {
+	log.Info("defaulting cronjob", "namespace", c.Namespace, "name", c.Name)
+
+	if c.Spec.ConcurrencyPolicy == "" {
+		c.Spec.ConcurrencyPolicy = AllowConcurrent
+	}
+	if c.Spec.Suspend == nil {
+		c.Spec.Suspend = new(bool)
+	}
+	if c.Spec.SuccessfulJobsHistoryLimit == nil {
+		c.Spec.SuccessfulJobsHistoryLimit = new(int32)
+		*c.Spec.SuccessfulJobsHistoryLimit = 3
+	}
+	if c.Spec.FailedJobsHistoryLimit == nil {
+		c.Spec.FailedJobsHistoryLimit = new(int32)
+		*c.Spec.FailedJobsHistoryLimit = 1
+	}
+}
+
+/*
+We use the `webhook.Validator` interface to validate our CRD.
+A webhook will automatically be served that calls the validation.
+*/
+
+var _ webhook.Validator = &CronJob{}
+
+/*
+The ValidateCreate method is expected to validate that its receiver upon creation.
+If not, it return an error that's used as a validation message.
+We separate out ValidateCreate from ValidateUpdate to allow behavior
+like making certain fields immutable, so that they can only be set on creation.
+Here, however, we just use the same shared validation.
+*/
+func (c *CronJob) ValidateCreate() error {
+	log.Info("validate create", "namespace", c.Namespace, "name", c.Name)
+	return c.validateCronJob()
+}
+
+/*
+The ValidateUpdate method is expected to validate the receiver upon update.
+If not, it return an error that's used as a validation message.
+The receiver is the new object and the argument is the old object.
+*/
+func (c *CronJob) ValidateUpdate(old runtime.Object) error {
+	log.Info("validate update", "namespace", c.Namespace, "name", c.Name)
+	return c.validateCronJob()
+}
+
+/*
+We validate the name and the spec of the CronJob.
+*/
+
+func (c *CronJob) validateCronJob() error {
+	allErrs := c.validateCronJobSpec()
+	if err := c.validateCronJobName(); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "batch.tutorial.kubebuilder.io", Kind: "CronJob"},
+		c.Name, allErrs)
+}
+
+/*
+Some fields are declaratively validated by openapi schema.
+You can find kubebuilder validation markers (prefixed
+with `// +kubebuilder:validation`) in the [API](api-design.md)
+You can find all of the kubebuilder supported markers for
+declaring validation in [here](../TODO.md).
+*/
+
+func (c *CronJob) validateCronJobSpec() field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	spec := c.Spec
+	fldPath := field.NewPath("spec")
+
+	if len(spec.Schedule) == 0 {
```

Comment:
perhaps put an example of actually parsing the schedule and seeing if it has an error?